lib/src/priority_queue.dart

```diff
@@ -418,3 +445,41 @@ class HeapPriorityQueue<E> implements PriorityQueue<E> {
     _queue = newQueue;
   }
 }
+
+/// Implementation of [HeapPriorityQueue.unorderedElements].
+class _UnorderedElementsIterable<E> extends Iterable<E> {
+  final HeapPriorityQueue<E> _queue;
+  _UnorderedElementsIterable(this._queue);
+  @override
+  Iterator<E> get iterator => _UnorderedElementsIterator<E>(_queue);
```

Comment:
Can we simply use _queue._queue.iterator and then use the concurrent modification checks already done by the List class? I guess not, because of removeAll() and maybe other operations (like adding a duplicate element) won't properly catch concurrent modifications.

Reply:
We can't. The only thing `list.iterator` checks for is length changes. A list is treated like a map from integers to values, and the only structural change is changing which integers are allowed - the length. Changing values is like changing the values of a `Map`, it's not something which affects iteration. (Corollary: a fixed length list doesn't need to check for concurrent modifications).

Also, the `_queue` backing store is a `List<E?>` which may be longer than the number of elements. We need to restrict it to the first `length` elements and cast it to `E`. 
A `_queue.take(length).cast<E>()` would do it, but less efficiently, and wouldn't catch most modifications.

The modifications to worry about are those which causes an iteration to repeat or skip an element which was in the queue during the entire iteration. That's what I consider an *inconsistent* iteration. 
Not emitting an element which was removed or added during iteration is not a problem. Emitting an element after it was removed is not great, but acceptable - it was part of the collection that was originally asked to be iterated.
For a queue like this, most operations change the position of elements, which means that they risk duplicating or skipping elements when we are iterating by index. A `removeAll` is not actually a problem, we could allow iteration to continue (and then immediately stop because there is no next element), but operations adding or removing single elements are.

I originally wrote
```dart
Iterable<E> get unorderedElements sync* {
  var modificationCount = _modificationCount;
  for (var i = 0; i < _length; i++) {
    yield _queue[i] ?? null as E;
    if (modificationCount != _modificationCount) {
      throw ConcurrentModificationError(this);
    }
  }
}
```
but I'm assuming that `sync*` implementations will probably also be less efficient than a hand-written implementation (in both code size and performance).

Reply:
Oh wow I didn't need that thorough an answer but that was interesting to read.

Reply:
I don't like to just say "no" like Terkel's dad :grin:.